### PR TITLE
[zfp] update to 1.0.0

### LINKED ISSUE #28574: [zfp] update to 1.0
Library name: ZFP

New version number: 1.0

It was released some time ago the 1.0 version of the zfp package: https://github.com/LLNL/zfp/releases/tag/1.0.0

### PR DESCRIPTION
**Pull request by JonLiu1993**
Fix #28574 

Update zfp to the latest version 1.0.0.
Modernize the portfile.cmake file

All features are tested successfully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

### CODE REVIEW

File: ports/zfp/vcpkg.json [hunk 1]
```diff
@@ -1,9 +1,19 @@
 {
   "name": "zfp",
-  "version-string": "0.5.5",
-  "port-version": 3,
+  "version": "0.5.5",
```
Comment (dg0yt) [inline ports/zfp/vcpkg.json:3]: You missed to update this to 1.0.0...

File: versions/z-/zfp.json [hunk 2]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "d562c703fcb471ea4eb29330ce91736c6cb588cb",
+      "version": "0.5.5",
+      "port-version": 0
+    },
```
Comment (dg0yt) [inline versions/z-/zfp.json:7]: ... and now we have a strange sequence here.

File: ports/zfp/portfile.cmake [hunk 3]
```diff
@@ -1,44 +1,46 @@
...
-vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/${PORT})
+vcpkg_cmake_config_fixup(CONFIG_PATH lib/cmake/${PORT})
 
 # Rename problematic root include "bitstream.h"; conflicts with x265's private headers
-file(RENAME ${CURRENT_PACKAGES_DIR}/include/bitstream.h ${CURRENT_PACKAGES_DIR}/include/zfp/bitstream.h)
 vcpkg_replace_string(${CURRENT_PACKAGES_DIR}/include/zfp.h "\"bitstream.h\"" "\"zfp/bitstream.h\"")
 
+if("cfp" IN_LIST FEATURES AND VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+    vcpkg_copy_tools(TOOL_NAMES zfpcmd AUTO_CLEAN)
+endif()
```
Comment (dg0yt) [inline ports/zfp/portfile.cmake:42]: I assume the tool exists also for static builds. But the `bin` dir is already removed at this point.

Doesn't  the utility feature build any tools?

File: ports/zfp/portfile.cmake [hunk 4]
```diff
@@ -1,44 +1,46 @@
...
-vcpkg_install_cmake()
+vcpkg_cmake_install()
 
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
 
 if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
-    file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin ${CURRENT_PACKAGES_DIR}/debug/bin)
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/bin" "${CURRENT_PACKAGES_DIR}/debug/bin")
 endif()
```
Comment (dg0yt) [inline ports/zfp/portfile.cmake:33]: This where we loose the tools for static builds. This explicitly removal shouldn't be needed if we use `vcpkg_copy_tools(... AUTO_CLEAN)`  correctly. However, we may need to construct a list of tools depending on actual features.

**Review by dg0yt — CHANGES_REQUESTED** [review]
I know it is a draft PR but I want to record some early comments.

File: ports/zfp/fix-install-tools.patch [hunk 5]
```diff
@@ -1,72 +1,77 @@
...
++install(TARGETS inplace 
++    RUNTIME DESTINATION tools)
++install(TARGETS iterator 
++    RUNTIME DESTINATION tools)
++install(TARGETS pgm 
++    RUNTIME DESTINATION tools)
++install(TARGETS simple 
++    RUNTIME DESTINATION tools)
++install(TARGETS speed 
++    RUNTIME DESTINATION tools)
```
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:50]: Multiple targets could be installed in a single command.

File: ports/zfp/portfile.cmake (hunk 3 above)
Comment (JonLiu1993) [reply ports/zfp/portfile.cmake:42]: `zfpcmd.exe` is built from utilities

**Comment by aavbsouza** [comment]
Any news on this issue ?

**Comment by MonicaLiu0311** [comment]
Ping @JonLiu1993

File: ports/zfp/fix-install-tools.patch [hunk 6]
```diff
@@ -1,72 +1,77 @@
...
+ option(BUILD_ALL "Build all subdirectories" OFF)
+ if(BUILD_ALL)
+   set(BUILD_CFP ON CACHE BOOL "Build CFP arrays library" FORCE)
+-  set(BUILD_ZFORP ON CACHE BOOL "Build Fortran library" FORCE)
+-  set(BUILD_ZFPY ON CACHE BOOL "Build python bindings for zfp" FORCE)
++  set(BUILD_ZFORP OFF CACHE BOOL "Build Fortran library" FORCE)
++  set(BUILD_ZFPY OFF CACHE BOOL "Build python bindings for zfp" FORCE)
+   set(BUILD_UTILITIES ON CACHE BOOL "Build command line utilities for zfp" FORCE)
+   set(BUILD_EXAMPLES ON CACHE BOOL "Build Examples" FORCE)
+ endif()
```
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:15]: Does have `BUILD_ALL` any other effects than these? If not, `BUILD_ALL` shouldn't be used, the patch chunk should be removed, and these options should be controlled directly be feature `all`.

File: ports/zfp/fix-install-tools.patch (hunk 5 above)
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:50]: So upstream doesn't install examples even if they are build?  Why does vcpkg have to install them?
Why not just drop examples in this case?

File: ports/zfp/fix-install-tools.patch [hunk 7]
```diff
@@ -1,72 +1,77 @@
...
+index d15246e..186bc4e 100644
+--- a/tests/CMakeLists.txt
++++ b/tests/CMakeLists.txt
+@@ -19,3 +19,6 @@ if(ZFP_BUILD_TESTING_LARGE)
+     endforeach()
+   endforeach()
+ endif()
++
++install(TARGETS testzfp 
++    RUNTIME DESTINATION tools)
```
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:62]: Upstream doesn't install tests. Does vcpkg really have to add it?

File: ports/zfp/fix-install-tools.patch [hunk 8]
```diff
@@ -1,72 +1,77 @@
...
+@@ -10,6 +10,9 @@ if(HAVE_LIBM_MATH)
+   target_link_libraries(zfpcmd m)
+ endif()
+ 
++install(TARGETS zfpcmd 
++    RUNTIME DESTINATION tools)
++
+ if(BUILD_UTILITIES)
+   install(TARGETS zfpcmd
+     DESTINATION "${CMAKE_INSTALL_BINDIR}")
```
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:77]: Isn't this utility installed twice here?

File: ports/zfp/fix-install-tools.patch [hunk 9]
```diff
@@ -1,72 +1,77 @@
...
+diff --git a/cfp/CMakeLists.txt b/cfp/CMakeLists.txt
+index 3d8af6e..b669601 100644
+--- a/cfp/CMakeLists.txt
++++ b/cfp/CMakeLists.txt
+@@ -34,3 +34,5 @@ install(TARGETS cfp EXPORT cfp-targets
+   RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+   LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+   ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
++  
++install(DIRECTORY ${ZFP_SOURCE_DIR}/include DESTINATION include)
```
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:25]: zfp already has installation of public headers:
https://github.com/LLNL/zfp/blob/f39af72648a2aeb88e9b2cca8c64f51b493ad5f4/CMakeLists.txt#L329-L334

**Review by dg0yt — CHANGES_REQUESTED** [review]
Maybe the 1.0.0. release is an opportunity to trim the port's patching?

File: ports/zfp/fix-install-tools.patch (hunk 6 above)
Comment (JonLiu1993) [reply ports/zfp/fix-install-tools.patch:15]: `BUILD_ALL `is used as a macro judgment in the upstream Config file.
https://github.com/LLNL/zfp/blob/81dbdd02dc3d326b0565336c378c526263eba6ac/Config#L82-L91

Comment (dg0yt) [reply ports/zfp/fix-install-tools.patch]: `Config` is part of the `Makefile` build system, not of the CMake build system. So this spot is unrelated.

File: ports/zfp/fix-install-tools.patch [hunk 10]
```diff
@@ -11,62 +11,8 @@ index 47179fd..96add54 100644
 +  set(BUILD_ZFORP OFF CACHE BOOL "Build Fortran library" FORCE)
 +  set(BUILD_ZFPY OFF CACHE BOOL "Build python bindings for zfp" FORCE)
    set(BUILD_UTILITIES ON CACHE BOOL "Build command line utilities for zfp" FORCE)
-   set(BUILD_EXAMPLES ON CACHE BOOL "Build Examples" FORCE)
+-  set(BUILD_EXAMPLES ON CACHE BOOL "Build Examples" FORCE)
++  set(BUILD_EXAMPLES OFF CACHE BOOL "Build Examples" FORCE)
```
Comment (dg0yt) [inline ports/zfp/fix-install-tools.patch:15]: I still think we should get rid this patch and of `BUILD_ALL`. We can control the desired options directly if `BUILD_ALL`  is not used.

Comment (JonLiu1993) [reply ports/zfp/fix-install-tools.patch]: Thanks for your review, done.

File: ports/zfp/vcpkg.json [hunk 11]
```diff
@@ -15,9 +15,6 @@
     }
   ],
   "features": {
-    "all": {
-      "description": "Build all components"
-    },
```
Comment (dg0yt) [inline ports/zfp/vcpkg.json:10]: Removing features is still somewhat tricky in vcpkg, in particular breaking the `upgrade` command.
Simply leave the feature, make it depend on features cfp, utility.
Change the description to `Obsolete. This feature is left for compatibility.`
(Similar to libmariadb[zlib].)

Comment (JonLiu1993) [reply ports/zfp/vcpkg.json]: Thanks for your review, Done.

**Review by MonicaLiu0311 — APPROVED** [review]

**Pull request merged**